lib/Sema/SemaExpr.cpp

```diff
@@ -11935,7 +11935,20 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
                                       CountBoundsExpr::Kind Kind,
                                       Expr *CountExpr,
                                       SourceLocation RParenLoc) {
-  return new (Context) CountBoundsExpr(Kind, CountExpr, BoundsKWLoc, RParenLoc);
+  ExprResult Result = UsualUnaryConversions(CountExpr);
+  if (Result.isInvalid())
+    return ExprError();
+  Expr *PromotedCountExpr = Result.get();
+  QualType ResultType = PromotedCountExpr->getType();
+  if (!ResultType->isIntegerType()) {
```

Comment:
The spec allows the integer to be signed or unsigned (it has to be an integral type).

Interestingly, the C definition itself is silent on what happens to pointer arithmetic if you add a pointer type and an integer whose range is larger than that of pointer types.  The definition simply requires that the operand have integer type.   In the case of Checked C, we require that it be provable that count expressions not cause an overflow when they are scaled by the size of the pointer referent type.  This implies that the value must fit into a pointer-sized integer.   This doesn't need to part of typechecking.  We still would need to make sure compilers do something reasonable when asked to add a 64-bit integer to a 32-bit pointer type.